Framework/Beamline/src/DetectorInfo.cpp

```diff
@@ -78,28 +90,250 @@ bool DetectorInfo::isEquivalent(const DetectorInfo &other) const {
+ * incremented by the scan count of that detector in `this`. The relative order
+ * of time indices added from `other` is preserved. If the interval for a time
+ * index in `other` is identical to a corresponding interval in `this`, it is
+ * ignored, i.e., no time index is added. */
+void DetectorInfo::merge(const DetectorInfo &other) {
+  if (size() != other.size())
+    failMerge("size mismatch");
+  if (!m_scanIntervals || !other.m_scanIntervals)
+    failMerge("scan intervals not defined");
+  if (!(m_isMonitor == other.m_isMonitor) &&
+      (*m_isMonitor != *other.m_isMonitor))
+    failMerge("monitor flags mismatch");
+  // TODO If we make masking time-independent we need to check masking here.
+
+  std::vector<bool> merge(other.m_positions->size(), true);
+
+  for (size_t linearIndex1 = 0; linearIndex1 < other.m_positions->size();
+       ++linearIndex1) {
+    const size_t detIndex = getIndex(other.m_indices, linearIndex1).first;
+    const auto &interval1 = (*other.m_scanIntervals)[linearIndex1];
+    for (size_t timeIndex = 0; timeIndex < scanCount(detIndex); ++timeIndex) {
+      const auto linearIndex2 = linearIndex({detIndex, timeIndex});
+      const auto &interval2 = (*m_scanIntervals)[linearIndex2];
+      if (interval1 == interval2) {
+        if ((*m_isMasked)[linearIndex2] != (*other.m_isMasked)[linearIndex1])
+          failMerge("matching scan interval but mask flags differ");
+        if ((*m_positions)[linearIndex2] != (*other.m_positions)[linearIndex1])
+          failMerge("matching scan interval but positions differ");
+        if ((*m_rotations)[linearIndex2].coeffs() !=
+            (*other.m_rotations)[linearIndex1].coeffs())
+          failMerge("matching scan interval but rotations differ");
+        merge[linearIndex1] = false;
+      } else if ((interval1.first < interval2.second) &&
+                 (interval1.second > interval2.first)) {
+        failMerge("scan intervals overlap but not identical");
+      }
+    }
+  }
+
+  if (!m_scanCounts)
+    initScanCounts();
+  if (!m_indexMap)
+    initIndices();
+  // Temporary to accumulate scan counts (need original for index offset).
+  auto scanCounts(m_scanCounts);
+  for (size_t linearIndex = 0; linearIndex < other.m_positions->size();
+       ++linearIndex) {
+    if (!merge[linearIndex])
+      continue;
+    auto newIndex = getIndex(other.m_indices, linearIndex);
+    const size_t detIndex = newIndex.first;
+    newIndex.second += scanCount(detIndex);
+    scanCounts.access()[detIndex]++;
+    m_indexMap.access()[detIndex].push_back((*m_indices).size());
+    m_indices.access().push_back(newIndex);
+    m_isMasked.access().push_back((*other.m_isMasked)[linearIndex]);
+    m_positions.access().push_back((*other.m_positions)[linearIndex]);
+    m_rotations.access().push_back((*other.m_rotations)[linearIndex]);
+    m_scanIntervals.access().push_back((*other.m_scanIntervals)[linearIndex]);
+  }
+  m_scanCounts = std::move(scanCounts);
+}
```

Comment:
I feel that reserving would be irrelevant. Given that `std::vector` typically(?) doubles its size when it grows I would expect that we have typically have at most one single reallocation per merge. If you merge many objects you might even do worse with `reserve`, since it might cause *more* reallocations.